### PR TITLE
task6-review

### PR DESCRIPTION
**Pull request by duartenunes86**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ app.use((err, request, response, next) =>{
     response.status(404).send(err.msg)
```
Comment (aloush-dev) [inline app.js]: we could make this more dynamic and instead of saying if err.status===404 and just say if err.status and err.msg are true.

Then we can send err.status and err.msg back to the client

**Review by aloush-dev — CHANGES_REQUESTED** [review]
On the whole really nice job.
There is going to be a lot of feedback here but it's because of the mess we currently have with branches so going forward feedback won't be this daunting.


Task 3 - Get /api
For you endpoint of /api you are still going an fs.readfile in a model file, a model is strcitly for interacting with a database which you are not, the logic here is fine you just need to move it into your getAPI controller.

For your testing of this endpoint you are doing an fs readfile which is technically fine but rememeber you could just require the endpoints file in at the top of your test file to test against.

Task 4 - GET /api/articles/:article_id

Your first test could be much better tested in terms of the expects you are using look into using expect.objectContaining or .toMatchObject as they are a much more flexible matcher.
You are also testing against an array and this task should just respond with an object.

I would like to see another test for a 400, what if I put a string in there of 'banana' instead of an ID number? it should return a 400.
Just a note PSQL will throw an error here for you asnd you can deal with it in your error handlers.

In your controller if you're going to test the length of the article array which you get back you should do this in the model and reject your promise from there.

It seems like you have a promise reject in both the model and controller which are doing the same thing we don't need the one in the controller.

It's time to also start thinking about splitting your error handling middleware into seperate functions for each usecase, PSQL Error, Custom Errors for example.

Task 5 - GET /api/articles

Nice first test again as same above look into expect.objectContaining or .toMatchObject .
Also it doesn't need to be a seperate test but add to your description that you are excluding the body from the object you are returning.

We are also not testing that the articles come back in any sort of order. Look into the package Jest sorted to test this.

Nice model you have ordered by correctly.

As with the other task your logic for handling the 404 error is fine just move it over to the model instead of the controller.

The res.status.send in your catch should really be in an error handler in your app.js so that it can be caught for all errors and not just this promise chain.

Task - 6 GET /api/articles/:article_id/comments

Testing wise nice job see other comments about better matchers.

Add a test for an invalid ID for example if I send a string and not a number.

I would also like to see a test for a status 200 valid ID but if there is no comments? we should still get an empty array as this is not an error

In your controller I would send the comments back on an appropriate key such as comments instead of the key of results.

In your model you have a promise reject which says the article doesn't exist if the rows array is falsey but thats not true as it might be a valid article but with no comments attached in which we should just return the empty array., this same promise reject exists in the controller.


I know this is a lot of feedback but please implement this feedback on this branch before continuing any further. Task wise you are perfectly on track and not behind so take the time to sort these issues out so that going forward we won't have these branch issues and we can provide feedback on a ticket by ticket basis.

File: models/articles-models.js [hunk 2]
```diff
@@ -1,12 +1,18 @@
 const db = require('../db/connection');
 
 exports.selectArticleById = (id) => {
+  if(isNaN(id)){return Promise.reject({status:400, msg:"Invalid id" })}
```
Comment (nc-poppy-m) [inline models/articles-models.js]: This case is dealt with for you by PSQL, you don't need this line to check if id is a number

File: controllers/articles-controllers.js [hunk 3]
```diff
@@ -13,14 +13,20 @@ const {
...
+      res.status(200).send(  {article });
+
 
-      res.status(200).send( {article });
 
-   } ) .catch((err) => {
-      next(err);
-    });}
+  }) 
+    .catch((err) => {
```
Comment (nc-poppy-m) [inline controllers/articles-controllers.js:23]: You should be passing your error to next so it can be passed to the error handling middleware you wrote - you were doing this correctly before :)

File: controllers/articles-controllers.js [hunk 4]
```diff
@@ -29,23 +35,23 @@ const {
           
           res.status(200).send( {results });
           
-        }) .catch((err) => {
-          next(err);
-        });
+        })
```
Comment (nc-poppy-m) [inline controllers/articles-controllers.js:36]: Comment your catch block back in

File: models/articles-models.js [hunk 5]
```diff
@@ -21,21 +27,33 @@ exports.selectArticles = () => {
...
+      } 
+      else{
       return result.rows;
+      }
     });
   };
 
   exports.selectCommentsByArticle = (id) =>{
-    return db.query(` SELECT   comment_id,
+    if(isNaN(id)){return Promise.reject({status:400, msg:"Invalid id" })}
```
Comment (nc-poppy-m) [inline models/articles-models.js]: Same thing, PSQL will send back an error if id is not a number, so you don't need to check for it here

File: controllers/articles-controllers.js [hunk 6]
```diff
@@ -29,23 +35,23 @@ const {
...
-          }
+          
             
           }) .catch((error) => {
-            res.status(500).send({ msg: 'An error occurred' });
-  });
-          };
-          
\ No newline at end of file
+            const{status, msg}=error
```
Comment (nc-poppy-m) [inline controllers/articles-controllers.js]: Just need to pass error to next in the error handling middleware chain using ```next``` , we don't do the res.status.send for the error here as this is handled by the error handling middleware

File: controllers/articles-controllers.js [hunk 7]
```diff
@@ -29,23 +35,23 @@ const {
           
           res.status(200).send( {results });
```
Comment (nc-poppy-m) [inline controllers/articles-controllers.js]: send this back on a key of ```articles``` rather than results

File: controllers/articles-controllers.js [hunk 8]
```diff
@@ -29,23 +35,23 @@ const {
...
         exports.getCommentsByArticle = (req, res, next) => {
         const id =req.params.article_id;
         
         
           selectCommentsByArticle(id).then((results) => {
            
-            if (results.length===0) {res.status(404).send({ msg:'article doesn\'t exist' });
-          }else{
+            
             res.status(200).send( {results });
```
Comment (nc-poppy-m) [inline controllers/articles-controllers.js]: similar to above, send back the comments on a key of ```comments``` rather than ```results```

**Review by nc-poppy-m — CHANGES_REQUESTED** [review]
Brilliant work on this mammoth of a task Duarte, you've done a great job implementing the feedback you got yesterday. You have a few more things to go before you merge, but you are so close!

- Look at my comments about PSQL error handling, you shouldn't be checking if article_id is a number in the models as PSQL will do that for you and send that to your error handling middleware
- Your .catch block should only pass the error along to the next error handling middleware in the chain using ```next(err)``` - you were previously doing this right
- We want the articles and comments sent back on sensible keys of articles and comments rather than results, so make these changes to your files

Have a look at my comments, make those changes and resubmit the pr once that's done and your tests pass. 🙂

File: models/articles-models.js [hunk 9]
```diff
@@ -1,18 +1,31 @@
 const db = require('../db/connection');
+const format = require('pg-format');
+
+const checkExists = async (table, column, value) => {
```
Comment (nc-poppy-m) [inline models/articles-models.js]: I'd really like to see this written consistently - why have you used async/await here and .then further down?

**Review by nc-poppy-m — CHANGES_REQUESTED** [review]
1 more change and that's it as far as I'm concerned - refactor your checkExists function from aync/await to .then so it's consistent. As soon as you've done that, resend the pr and that should be done!

**Review by nc-poppy-m — APPROVED** [review]
Done! Well done for powering through this. All good, please merge when you're ready :)
